userland/src/sync_parameters.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <iostream>
+#include <string>
+#include <thread>
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <simple_msgs/AllBuiltinTypes.h>
+#include <simple_msgs/AllDynamicArrayTypes.h>
+#include <simple_msgs/AllPrimitiveTypes.h>
+#include <simple_msgs/AllStaticArrayTypes.h>
+#include <simple_msgs/Nested.h>
+#include <simple_msgs/String.h>
+#include <simple_msgs/Uint32.h>
+
+#include <userland_msgs/AddTwoInts.h>
+
+int main(int argc, char ** argv)
+{
+  rclcpp::init(argc, argv);
+
+  auto node = rclcpp::Node::make_shared("parameters");
+
+  auto sync_parameters_client = std::make_shared<rclcpp::parameter::SyncParametersClient>(node);
+
+  auto parameters = {
+    rclcpp::parameter::ParameterVariant("foo", int64_t(2)),
+    rclcpp::parameter::ParameterVariant("bar", std::string("hello")),
+    rclcpp::parameter::ParameterVariant("baz", double(1.45)),
+    rclcpp::parameter::ParameterVariant("foobar", true),
+  };
+
+  auto results1 = sync_parameters_client->set_parameters(parameters);
+  for (auto v : results1) {
+    if (!v.successful) {
+      std::cerr << v.reason << std::endl;
+    }
+  }
```

Comment:
I think it would be good in the example to show how to associate the setting result and the parameter you requested to set, such that the message might be `Failed to set parameter 'foo' as a 'type' with value 'value': 'reason'`. The reason I think it's good to introduce this into the example is it shows users how they might accomplish this and if it turns out to be awkward for us to do it, then maybe we need to rethink what we return from that function or we need to provide a helper function to make it easier.

For instance, we might provide a set parameters result "packing" or "zipping" function, which would look like this:

```c++
auto results = sync_parameters_client->set_parameters(parameters);
for (auto & result : rclcpp::parameters::pack_set_parameter_results(parameters, results)) {
  if (!result.successful) {
    std::cerr << "Failed to set parameter '" << result.parameter.name
              << "' as a '" << result.parameter.type
              << "' with value '" << result.parameter.to_string()
              << "': '" << result.reason << "'" << std::endl;
  }
}
```

What do you guys think?